simulators/eth2/engine/prepared_testnet.go

```diff
@@ -0,0 +1,260 @@
+		hivesim.Params{
+			"HIVE_CHECK_LIVE_PORT": "0",
+		},
+		consensusConfigOpts,
+	)
+
+	return &PreparedTestnet{
+		spec:          spec,
+		eth1Genesis:   eth1Genesis,
+		eth2Genesis:   state,
+		keys:          env.Secrets,
+		executionOpts: executionOpts,
+		beaconOpts:    beaconOpts,
+		validatorOpts: validatorOpts,
+		keyTranches:   keyTranches,
+	}
+}
+
+func (p *PreparedTestnet) createTestnet(t *hivesim.T) *Testnet {
+	genesisTime, _ := p.eth2Genesis.GenesisTime()
+	genesisValidatorsRoot, _ := p.eth2Genesis.GenesisValidatorsRoot()
+	return &Testnet{
+		t:                     t,
+		genesisTime:           genesisTime,
+		genesisValidatorsRoot: genesisValidatorsRoot,
+		spec:                  p.spec,
+		eth1Genesis:           p.eth1Genesis,
+	}
+}
+
+func (p *PreparedTestnet) startEth1Node(testnet *Testnet, eth1Def *hivesim.ClientDefinition, consensus ConsensusType) {
+	testnet.t.Logf("Starting eth1 node: %s (%s)", eth1Def.Name, eth1Def.Version)
+
+	opts := []hivesim.StartOption{p.executionOpts}
+	if len(testnet.eth1) == 0 {
+		// we only make the first eth1 node a miner
+		if consensus == Ethash {
+			opts = append(opts, hivesim.Params{"HIVE_MINER": "1212121212121212121212121212121212121212"})
+		} else if consensus == Clique {
+			opts = append(opts, hivesim.Params{
+				"HIVE_CLIQUE_PRIVATEKEY": "9c647b8b7c4e7c3490668fb6c11473619db80c93704c70893d3813af4090c39c",
+				"HIVE_MINER":             "658bdf435d810c91414ec09147daa6db62406379",
+			})
+		}
+	} else {
+		bootnode, err := testnet.eth1[0].EnodeURL()
+		if err != nil {
+			testnet.t.Fatalf("failed to get eth1 bootnode URL: %v", err)
+		}
+
+		// Make the client connect to the first eth1 node, as a bootnode for the eth1 net
+		opts = append(opts, hivesim.Params{"HIVE_BOOTNODE": bootnode})
+	}
+	en := &Eth1Node{testnet.t.StartClient(eth1Def.Name, opts...)}
+	dest, _ := en.EngineRPCAddress()
+	testnet.eth1 = append(testnet.eth1, en)
+	simIP, err := testnet.t.Sim.ContainerNetworkIP(testnet.t.SuiteID, "bridge", "simulation")
+	if err != nil {
+		panic(err)
+	}
+	secret, err := hex.DecodeString("7365637265747365637265747365637265747365637265747365637265747365")
+	if err != nil {
+		panic(err)
+	}
+	testnet.proxies = append(testnet.proxies, NewProxy(net.ParseIP(simIP), PortEngineRPC+len(testnet.eth1), dest, secret))
+}
+
+func (p *PreparedTestnet) startBeaconNode(testnet *Testnet, beaconDef *hivesim.ClientDefinition, eth1Endpoints []int) {
+	testnet.t.Logf("Starting beacon node: %s (%s)", beaconDef.Name, beaconDef.Version)
+
+	opts := []hivesim.StartOption{p.beaconOpts}
+	// Hook up beacon node to (maybe multiple) eth1 nodes
+	for _, index := range eth1Endpoints {
+		if index < 0 || index >= len(testnet.eth1) {
+			testnet.t.Fatalf("only have %d eth1 nodes, cannot find index %d for BN", len(testnet.eth1), index)
+		}
+	}
+
+	var addrs []string
+	var engineAddrs []string
+	for _, index := range eth1Endpoints {
+		eth1Node := testnet.proxies[index]
+		userRPC, err := eth1Node.UserRPCAddress()
```

Comment:
I had trouble using the `UserRPCAddress` of the proxy, I kept getting "the method net_version does not exist/is not available" errors.
I used `testnet.eth1[index].UserRPCAddress` for the `userRPC` and `testnet.proxies[index]` for `engineRPC`.